packages/filebrowser-extension/src/index.ts

```diff
@@ -273,18 +279,80 @@ function addCommands(app: JupyterLab, fbWidget: FileBrowser, docManager: IDocume
     isEnabled,
     execute: () => {
       if (isEnabled()) {
-        let context = docManager.contextForWidget(app.shell.currentWidget);
+        let context = docManager.contextForWidget(shell.currentWidget);
         return context.saveAs().then(() => {
           return context.createCheckpoint();
         });
       }
     }
   });
 
+  function getOpenArgs(args: JSONObject): { path: string, factory: string } {
+      let path = args['path'] as string || void 0;
+      let factory = args['factory'] as string || void 0;
+      return { path, factory };
+  }
+
+  function factoryMatch(path: string, factory='default'): boolean {
+    let ext = PathExt.extname(path);
+    if (!ext && factory !== 'default') {
+      return false;
+    }
+    if (factory === 'default') {
+      return  fbWidget.registry.defaultWidgetFactory(ext) !== void 0;
+    }
+    let factories = fbWidget.registry.preferredWidgetFactories(ext);
+    return find(factories, item => item.name === factory) !== undefined;
+  }
+
   commands.addCommand(CommandIDs.open, {
+    icon: args => {
+      // Only show the icon if no factory is given.
+      let { factory } = getOpenArgs(args);
+      return factory ? '' : 'jp-MaterialIcon jp-OpenFolderIcon';
+    },
+    mnemonic: args => {
+      // Only use the mnemonic if no factory is given.
+      let { factory } = getOpenArgs(args);
+      return factory ? -1 : 0;
+    },
+    label: args => {
+      // Show the factory as the label if given or use "Open".
+      let { factory } = getOpenArgs(args);
+      return factory || 'Open';
+    },
+    isEnabled: args => {
+      let { path, factory } = getOpenArgs(args);
+      // Check for an exact match.
+      if (path && factory) {
+        return factoryMatch(path, factory);
+      }
+      let match = true;
+      each(fbWidget.selection(), item => {
```

Comment:
not exactly following the semantics of this loop, can it can probably be replaced by `some` or `every`